src/masternodeman.cpp

```diff
@@ -705,6 +730,420 @@ void CMasternodeMan::ProcessMessage(CNode* pfrom, std::string& strCommand, CData
+            LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Found self at rank %d/%d, verifying up to %d masternodes\n",
+                        nMyRank, nRanksTotal, nCountMax);
+            break;
+        }
+        ++it;
+    }
+
+    // edge case: list is too short and this masternode is not enabled
+    if(nMyRank == -1) return;
+
+    // send verify requests to up to nCountMax masternodes starting from
+    // (MAX_POSE_RANK + nCountMax * (nMyRank - 1) + 1)
+    int nOffset = MAX_POSE_RANK + nCountMax * (nMyRank - 1);
+    if(nOffset >= (int)vecMasternodeRanks.size()) return;
+
+    std::vector<CMasternode*> vSortedByAddr;
+    BOOST_FOREACH(CMasternode& mn, vMasternodes) {
+        vSortedByAddr.push_back(&mn);
+    }
+
+    sort(vSortedByAddr.begin(), vSortedByAddr.end(), CompareByAddr());
+
+    it = vecMasternodeRanks.begin() + nOffset;
+    while(it != vecMasternodeRanks.end()) {
+        if(it->second.IsPoSeVerified() || it->second.IsPoSeBanned()) {
+            LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Already %s%s%s masternode %s address %s, skipping...\n",
+                        it->second.IsPoSeVerified() ? "verified" : "",
+                        it->second.IsPoSeVerified() && it->second.IsPoSeBanned() ? " and " : "",
+                        it->second.IsPoSeBanned() ? "banned" : "",
+                        it->second.vin.prevout.ToStringShort(), it->second.addr.ToString());
+            ++it;
+            continue;
+        }
+        LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Verifying masternode %s rank %d/%d address %s\n",
+                    it->second.vin.prevout.ToStringShort(), it->first, nRanksTotal, it->second.addr.ToString());
+        if(SendVerifyRequest((CAddress)it->second.addr, vSortedByAddr)) {
+            nCount++;
+            if(nCount >= nCountMax) break;
+        }
+        ++it;
+    }
+
+    LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Sent verification requests to %d masternodes\n", nCount);
+}
+
+void CMasternodeMan::CheckSameAddr()
+{
+    if(!masternodeSync.IsSynced() || vMasternodes.empty()) return;
+
+    std::vector<CMasternode*> vBan;
+    std::vector<CMasternode*> vSortedByAddr;
+
+    {
+        LOCK(cs);
+
+        CMasternode* pprevMasternode = NULL;
+        CMasternode* pverifiedMasternode = NULL;
+
+        BOOST_FOREACH(CMasternode& mn, vMasternodes) {
+            vSortedByAddr.push_back(&mn);
+        }
+
+        sort(vSortedByAddr.begin(), vSortedByAddr.end(), CompareByAddr());
+
+        BOOST_FOREACH(CMasternode* pmn, vSortedByAddr) {
+            // check only (pre)enabled masternodes
+            if(!pmn->IsEnabled() && !pmn->IsPreEnabled()) continue;
+            // initial step
+            if(!pprevMasternode) {
+                pprevMasternode = pmn;
+                continue;
+            }
+            // second+ step
+            if(pmn->addr == pprevMasternode->addr) {
```

Comment:
This logic doesn't look quite right.  Suppose the first mn with a given address is correct.  Since it's address differs from that of the previous mn (assume this is not the first address in the list) then the outer condition is not satisfied and pverifiedMasternode doesn't get set.  Then the second, incorrect, mn doesn't get banned.